tfx/components/transform/executor.py

```diff
@@ -350,27 +349,38 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
 
     analyze_data_paths = []
     for split in splits_config.analyze:
-      data_uri = artifact_utils.get_split_uri(input_dict[EXAMPLES_KEY], split)
-      analyze_data_paths.append(io_utils.all_files_pattern(data_uri))
+      data_uris = artifact_utils.get_split_uris(input_dict[EXAMPLES_KEY], split)
+      if len(data_uris) != len(input_dict[EXAMPLES_KEY]):
+        raise ValueError(
+            'Analyze split does not exist over all example artifacts: %s' %
+            split)
+      for data_uri in data_uris:
+        analyze_data_paths.append(io_utils.all_files_pattern(data_uri))
 
     transform_data_paths = []
     materialize_output_paths = []
     if output_dict.get(TRANSFORMED_EXAMPLES_KEY) is not None:
-      transformed_example_artifact = artifact_utils.get_single_instance(
-          output_dict[TRANSFORMED_EXAMPLES_KEY])
-      # TODO(b/161490287): move the split_names setting to executor for all
-      # components.
-      transformed_example_artifact.split_names = (
-          artifact_utils.encode_split_names(list(splits_config.transform)))
+      for transformed_example_artifact in output_dict[TRANSFORMED_EXAMPLES_KEY]:
+        # TODO(b/161490287): move the split_names setting to executor for all
+        # components.
+        transformed_example_artifact.split_names = (
+            artifact_utils.encode_split_names(list(splits_config.transform)))
 
       for split in splits_config.transform:
-        data_uri = artifact_utils.get_split_uri(input_dict[EXAMPLES_KEY], split)
-        transform_data_paths.append(io_utils.all_files_pattern(data_uri))
-        transformed_example = os.path.join(
-            artifact_utils.get_split_uri(output_dict[TRANSFORMED_EXAMPLES_KEY],
-                                         split),
-            _DEFAULT_TRANSFORMED_EXAMPLES_PREFIX)
-        materialize_output_paths.append(transformed_example)
+        data_uris = artifact_utils.get_split_uris(input_dict[EXAMPLES_KEY],
+                                                  split)
+        if len(data_uris) != len(input_dict[EXAMPLES_KEY]):
+          raise ValueError(
+              'Transform split does not exist over all example artifacts: %s' %
+              split)
+        for data_uri in data_uris:
+          transform_data_paths.append(io_utils.all_files_pattern(data_uri))
+
+        transformed_example_uris = artifact_utils.get_split_uris(
+            output_dict[TRANSFORMED_EXAMPLES_KEY], split)
+        for output_uri in transformed_example_uris:
+          materialize_output_paths.append(
+              os.path.join(output_uri, _DEFAULT_TRANSFORMED_EXAMPLES_PREFIX))
```

Comment:
just confirm, can this makes sure one input artifact maps to one output artifact?

e.g.,

artifact1:train_split -> output artifact1:train_split
artifact1:eval_split -> output artifact1:eval_split
artifact2:train_split -> output artifact2: train_split
artifact2:eval_split -> output artifact2: eval_split

it would be good to have a test that input split has different row count, and verify output split row count

Reply:
i was thinking of just loading in the non-transformed files and comparing the count to the transformed files. is that what you're suggsting, or maybe something else?

Reply:
updated tests. new behavior is using a temp directory for testdata and creating two artifact directories (one with double the data of the other to more easily verify 1-to-1 mapping). in `verify_transform_output` for each corresponding pair of input and output artifacts, we verify that the number of records in a certain split of the input artifact is equal to the number of records in the same split of the output transformed artifact.

Reply:
TODO(jjma): test Transform API if uri count is not matched for input and output

Reply:
https://github.com/tensorflow/tfx/blob/ff6917997340401570d05a4d3ebd6e8ab5760495/tfx/components/transform/executor.py#L904
https://github.com/tensorflow/tfx/blob/ff6917997340401570d05a4d3ebd6e8ab5760495/tfx/components/transform/executor.py#L1323
re: our discussion during sync, `_MakeDatasetList()` checks that the number of patterns matches output paths